File: algorithm/problem/linklist/82. 删除排序链表中的重复元素 II.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* deleteDuplicates(ListNode* head) {
        // 与删除重复元素类似 但是需要记录一下 删除的上一个元素 判断3-4连的情况
        if (!head)
        {
            return head;
        }
        ListNode* dummy_node = new ListNode(0, head);
        ListNode* current = dummy_node;
        while (current->next && current->next->next)
        {
            if (current->next->val == current->next->next->val)
            {
                // 保留当前的值
                int v = current->next->val;
                while (current->next && current->next->val == v)
                {
                    current->next = current->next->next;
                }
            } else
            {
                current = current->next;
            }
        }
        return dummy_node->next;
    }
};

File: algorithm/problem/linklist/82. 删除排序链表中的重复元素 II.go
package main

func deleteDuplicates2(head *ListNode) *ListNode {
	if head == nil {
		return head
	}
	var dummy *ListNode = &ListNode{
		Val:  0,
		Next: head,
	}
	var current *ListNode = dummy
	for current.Next != nil && current.Next.Next != nil {
		if current.Next.Val == current.Next.Next.Val {
			// 说明遇到开始重复的元素 统一处理掉 直到没
			val := current.Next.Val
			for current.Next != nil && current.Next.Val == val {
				current.Next = current.Next.Next
			}
		} else {
			current = current.Next
		}
	}
	return dummy.Next
}

/*
解题思路:
	需要定义一个哑节点用来保存链表的头信息
	然后使用一个当前节点指向哑节点
	哑节点的next为处理节点的头节点
	当cur 节点next == cur.next.next 表示遇到重复元素
	然后记录重复元素 一致循环处理 知道遇到不同元素
	然后 将当前元素指向cur.next
	最后返回 dummy.next即可
*/
